File: Java_Fundamentals_January_2017/Java_Advanced/Exersices/03. Stacks/src/SimpleTextEditor.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class SimpleTextEditor {
    public static void main(String[] args) throws IOException {
        BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in));

        int count = Integer.parseInt(in.readLine());

        ArrayDeque<String> txtOperations = new ArrayDeque<>();

        for(int i = 0; i < count; i++){
            String[] command = in.readLine().split("\\s+");

            switch(command[0]){
                case "1": AppendText(txtOperations, command[1]); break;
                case "2": EraseElements(txtOperations, Integer.parseInt(command[1])); break;
                case "3": PrintElement(txtOperations, Integer.parseInt(command[1]) - 1); break;
                case "4": txtOperations.pop();
            }
        }
    }

    private static void PrintElement(ArrayDeque<String> txtOperations, int i) {
        if(!txtOperations.isEmpty()){
            String txt = txtOperations.peek();
            if(i >= 0 && i <= txt.length()){
                System.out.println(txt.charAt(i));
            }
        }
    }

    private static void EraseElements(ArrayDeque<String> txtOperations, int i) {
        if(!txtOperations.isEmpty()){
            String txt = txtOperations.peek();
            if(i <= txt.length()){
                txtOperations.push(txt.substring(0, txt.length() - i));
            }
        }
    }

    private static void AppendText(ArrayDeque<String> txtOperations, String s) {
        if(txtOperations.isEmpty()){
            txtOperations.push(s);
            return;
        }

        txtOperations.push(txtOperations.peek() + s);
    }
}

File: Java_Fundamentals_January_2017/Java_Advanced/Exersices/02. Matrices C#/Matrices/Simple Text Editor/SimpleTextEditor.cs
namespace Simple_Text_Editor
{
    using System;
    using System.Collections.Generic;

    public class SimpleTextEditor
    {
        public static void Main(string[] args)
        {
            int countOfOperations = int.Parse(Console.ReadLine());

            Stack<string> textOperations = new Stack<string>();

            for (int i = 0; i < countOfOperations; i++)
            {
                string[] command = Console.ReadLine().Split(' ');

                switch (command[0])
                {
                    case "1": AppendText(textOperations, command[1]);
                        break;
                    case "2":
                        int count = int.Parse(command[1]);
                        EraseElementsFromText(textOperations, count); break;
                    case "3":
                        int index = int.Parse(command[1]) - 1;
                        PrintElement(textOperations, index); break;
                    case "4":
                        textOperations.Pop(); break;
                }
            }

        }

        private static void PrintElement(Stack<string> textOperations, int index)
        {
            if (textOperations.Count != 0)
            {
                string text = textOperations.Peek();
                if (index < text.Length && index >= 0)
                {
                    Console.WriteLine(text[index]);
                }
            }
        }

        private static void EraseElementsFromText(Stack<string> textOperations, int count)
        {
            if (textOperations.Count != 0)
            {
                string text = textOperations.Peek();
                text = text.Substring(0, text.Length - count);
                textOperations.Push(text);
            }
        }

        private static void AppendText(Stack<string> textOperations, string s)
        {
            if (textOperations.Count == 0)
            {
                textOperations.Push(s);
                return;
            }

            textOperations.Push(textOperations.Peek() + s);
        }
    }
}
